be/src/runtime/buffer_control_block.cpp

```diff
@@ -107,19 +107,53 @@ Status BufferControlBlock::add_batch(TFetchDataResult* result) {
 
     if (_waiting_rpc.empty()) {
         _buffer_rows += num_rows;
-        _batch_queue.push_back(result);
+        _batch_queue.push_back(result.release());
         _data_arriaval.notify_one();
     } else {
         auto* ctx = _waiting_rpc.front();
         _waiting_rpc.pop_front();
-        ctx->on_data(result, _packet_num);
-        delete result;
+        ctx->on_data(result.get(), _packet_num);
         _packet_num++;
     }
     return Status::OK();
 }
 
-StatusOr<bool> BufferControlBlock::try_add_batch(TFetchDataResult* result) {
+Status BufferControlBlock::add_batch(std::vector<std::unique_ptr<TFetchDataResult>>& results) {
```

Comment:
Could this method share some code with the other `add_batch` ?

Reply:
> Could this method share some code with the other `add_batch` ?

I think it can, I add a new method `_process_batch_without_lock`.